File: backend/app/schemas/backtest.py
from pydantic import BaseModel, Field, AwareDatetime
from datetime import date
from enum import Enum
from typing import List, Optional, Union, Literal

class SmaCrossoverStrat(BaseModel):
    kind: Literal["sma_crossover"] = "sma_crossover"
    fast: int = Field(..., description="Fast SMA window length (bars).")
    slow: int = Field(..., description="Slow SMA window length (bars).")

Strategy = SmaCrossoverStrat

class Side(str, Enum):
    BUY = "buy"
    SELL = "sell"

class Timeframe(str, Enum):
    OneMin = "1m"
    FiveMin = "5m"
    OneHour = "1h"
    OneDay = "1d"

class BacktestRequest(BaseModel):
    symbol: str
    start: date
    end: date
    timeframe: Timeframe = Timeframe.OneMin
    initial_cash: float = 10000.0
    strategy: Strategy
    save: bool = False

class Trade(BaseModel):
    time: AwareDatetime
    side: Side
    price: float
    qty: float
    cash_after: float
    position_after: float
    equity_after: float

class EquityPoint(BaseModel):
    time: AwareDatetime
    equity: float

class Metrics(BaseModel):
    cagr : float
    max_drawdown : float

class BacktestResult(BaseModel):
    symbol: str
    start: date
    end: date
    timeframe: Timeframe
    initial_cash: float
    strategy: Strategy
    save: bool
    metrics: Metrics
    trades: List[Trade] = Field(default_factory=list)
    equity_curve: List[EquityPoint] = Field(default_factory=list)
    id: Optional[str] = None

File: frontend/src/types/backtest.ts
/**
 * These mirror the Pydantic schemas in backend/app/schemas/backtest.py.
 * The shape must match exactly — FastAPI serialises to JSON using those
 * schemas and we deserialise on this side using these types.
 */

export type Strategy = {
  kind: "sma_crossover";
  fast: number;
  slow: number;
};

export type BacktestRequest = {
  symbol: string;
  start: string;       // "YYYY-MM-DD"
  end: string;         // "YYYY-MM-DD"
  timeframe: "1m" | "5m" | "1h" | "1d";
  initial_cash: number;
  strategy: Strategy;
  save: boolean;
};

export type Side = "buy" | "sell";

export type Trade = {
  time: string;        // ISO datetime string
  side: Side;
  price: number;
  qty: number;
  cash_after: number;
  position_after: number;
  equity_after: number;
};

export type EquityPoint = {
  time: string;        // ISO datetime string
  equity: number;
};

export type Metrics = {
  cagr: number;        // e.g. 0.18 means 18% per year
  max_drawdown: number; // e.g. -0.25 means worst drop was 25%
};

export type BacktestResult = {
  symbol: string;
  start: string;
  end: string;
  timeframe: string;
  initial_cash: number;
  strategy: Strategy;
  metrics: Metrics;
  trades: Trade[];
  equity_curve: EquityPoint[];
  id: string | null;
};
